File: frontend/src/types/message.ts
export type MessageRole = 'user' | 'notify' | 'assistant' | 'system' | 'tool';

export interface GenerationInfo {
  duration_ms: number;  // 生成用时（毫秒）
  status: 'completed' | 'error' | 'stopped';  // 生成状态
  error_message?: string | null;  // 错误信息
  tokens_used?: number;  // 使用的token数
  usage_info?: UsageInfo | null;
}

export type TaskContextMode = 'attached' | 'detached';

export interface UsageInfo {
  input_tokens?: number;
  output_tokens?: number;
  total_tokens?: number;
  cached_tokens?: number;
  reasoning_tokens?: number;
  cache_creation_input_tokens?: number;
  cache_read_input_tokens?: number;
  source?: string;
  raw?: Record<string, unknown>;
}

export interface NodeUsage {
  turn_usage?: UsageInfo;
  branch_usage?: UsageInfo;
  active_context_usage?: UsageInfo;
  model_context_window?: number | null;
}

export interface Message {
  id: string;
  role: MessageRole;
  content: string;
  subtype?: string | null;
  node_id: string;
  parent_node_id?: string;
  parent_id?: string;
  model?: string;
  tokens_used?: number;
  branch_total_tokens?: number;
  branch_usage_info?: UsageInfo | null;
  context_usage?: NodeUsage | null;
  timestamp: number;
  generation_info?: GenerationInfo | null;  // 生成信息（仅助手消息有）
  name?: string;
  tool_calls?: any[];
  tool_call_id?: string;
  reasoning?: string;
  compact_metadata?: {
    trigger?: 'manual' | 'auto' | string;
    pre_tokens?: number;
    messages_to_keep?: number;
    last_pre_compact_message_id?: string;
    restored_files?: Array<{
      filename: string;
      content?: string;
      truncated?: boolean;
    }>;
  } | null;
  is_visible_in_transcript_only?: boolean;
  import_files?: Array<{
    filename: string;
  }>;
  image_refs?: Array<{
    filename: string;
    mime_type?: string;
  }>;
  tool_permission_mode?: ToolPermissionMode | null;
  task_context_mode?: TaskContextMode | null;
}

export interface SendMessageRequest {
  content: string;
  parent_node_id: string;
  focus_new_node?: boolean;
  model_id?: string;
  provider_id?: string;
  reasoning_effort?: string | null;
  thinking_enabled?: boolean | null;
  tool_permission_mode?: ToolPermissionMode;
  task_context_mode?: TaskContextMode;
  import_files?: Array<{
    filename: string;
  }>;
  image_refs?: Array<{
    filename: string;
    mime_type?: string;
  }>;
}

export type ToolPermissionMode = 'auto_approve' | 'modify_only' | 'ask_always' | 'plan';

File: backend/core/chat/message.py
# chat/message.py - 更新消息管理
from time import time
from typing import Optional, List
import uuid
from ..config.types import Message, Role

class MessageManager:
    """消息管理器"""
    
    @staticmethod
    def create_message(
        role: Role,
        content: str,
        parent_id: Optional[str] = None,
        name: Optional[str] = None,
        tool_calls: Optional[List[dict]] = None,
        tool_call_id: Optional[str] = None,
        model_id: Optional[str] = None
    ) -> Message:
        """创建消息"""
        message_id = str(uuid.uuid4())
        return {
            "id": message_id,
            "role": role,
            "content": content,
            "name": name,
            "tool_calls": tool_calls,
            "tool_call_id": tool_call_id,
            "timestamp": int(time()),
        }
    
    @staticmethod
    def create_system_message(content: str) -> Message:
        """创建系统消息"""
        return MessageManager.create_message(Role.SYSTEM, content)
    
    @staticmethod
    def create_user_message(content: str, parent_id: Optional[str] = None) -> Message:
        """创建用户消息"""
        return MessageManager.create_message(Role.USER, content, parent_id=parent_id)
    
    @staticmethod
    def create_assistant_message(
        content: str,
        parent_id: Optional[str] = None,
        tool_calls: Optional[List[dict]] = None
    ) -> Message:
        """创建助手消息"""
        return MessageManager.create_message(
            Role.ASSISTANT,
            content,
            parent_id=parent_id,
            tool_calls=tool_calls
        )
    
    @staticmethod
    def create_tool_message(content: str, tool_call_id: str, parent_id: Optional[str] = None) -> Message:
        """创建工具消息"""
        return MessageManager.create_message(
            Role.TOOL,
            content,
            parent_id=parent_id,
            tool_call_id=tool_call_id
        )